### PR TITLE
Add SharePoint Loader

### PR DESCRIPTION
**Pull request by netoferraz**
- Added a loader (`SharePointLoader`) that can pull documents (`pdf`, `docx`, `doc`) from the [SharePoint Document Library](https://support.microsoft.com/en-us/office/what-is-a-document-library-3b5976dd-65cf-4c9e-bf5a-713c10ca2872).
- Added a Base Loader (`O365BaseLoader`) to be used for all Loaders that use [O365](https://github.com/O365/python-o365) Package
- Code refactoring on `OneDriveLoader` to use the new `O365BaseLoader`.

### CODE REVIEW

File: langchain/document_loaders/onedrive.py [hunk 1]
```diff
@@ -59,66 +41,27 @@ def fetch_mime_types(self) -> Dict[str, str]:
...
-            account.authenticate()
-        return account
+    def _fetch_mime_types(self) -> Dict[str, str]:
+        """This method will give you a Dictionary where the supported file types
+        are the keys and their corresponding mime types are the values."""
+        file_types = _SupportedFileTypes(file_types=["doc", "docx", "pdf"])
+        file_mime_types = file_types.fetch_mime_types()
+        return file_mime_types
 
-    def _get_folder_from_path(self, drive: Type[Drive]) -> Union[Folder, Drive]:
```
Comment (netoferraz) [inline langchain/document_loaders/onedrive.py:115]: This method receives an instance of the class instead of the constructor.

File: langchain/document_loaders/onedrive.py [hunk 2]
```diff
@@ -2,40 +2,22 @@
...
 
 if TYPE_CHECKING:
-    from O365 import Account
     from O365.drive import Drive, Folder
 
 SCOPES = ["offline_access", "Files.Read.All"]
 logger = logging.getLogger(__name__)
 
 
-class _OneDriveSettings(BaseSettings):
```
Comment (netoferraz) [inline langchain/document_loaders/onedrive.py:25]: Moved to `O365BaseLoader`

File: langchain/document_loaders/onedrive.py [hunk 3]
```diff
@@ -59,66 +41,27 @@ def fetch_mime_types(self) -> Dict[str, str]:
...
-                    self.settings.client_id,
-                    self.settings.client_secret.get_secret_value(),
-                ),
-                scopes=SCOPES,
-                token_backend=token_backend,
-                **{"raise_http_errors": False},
-            )
-            # make the auth
-            account.authenticate()
-        return account
```
Comment (netoferraz) [inline langchain/document_loaders/onedrive.py:113]: Moved to `O365BaseLoader`

File: langchain/document_loaders/onedrive.py [hunk 4]
```diff
@@ -145,68 +88,6 @@ def _get_folder_from_path(self, drive: Type[Drive]) -> Union[Folder, Drive]:
...
-                        "There isn't a file with"
-                        f"object_id {object_id} in drive {drive}."
-                    )
-                    continue
-                if file.is_file:
-                    if file.mime_type in list(file_mime_types.values()):
-                        loader = OneDriveFileLoader(file=file)
-                        docs.extend(loader.load())
-        return docs
-
```
Comment (netoferraz) [inline langchain/document_loaders/onedrive.py:209]: Moved to `O365BaseLoader`

File: langchain/document_loaders/onedrive_file.py [hunk 5]
```diff
@@ -16,15 +16,13 @@
 
 
 class OneDriveFileLoader(BaseLoader, BaseModel):
-    file: File = Field(...)
-
     class Config:
         arbitrary_types_allowed = True
 
-    def load(self) -> List[Document]:
+    def load(self, file: File) -> List[Document]:  # type: ignore
```
Comment (netoferraz) [inline langchain/document_loaders/onedrive_file.py:22]: I had to change the function signature here because I got a `pydantic` error related to circular references. I've tried to use the method `.update_forward_refs()` but haven't succeeded.

```python
Traceback (most recent call last):
  File "o365_test.py", line 14, in <module>
    documents = loader.load()
  File "/home/netoferraz/opensource/langchain-o365/langchain/document_loaders/onedrive.py", line 172, in load
    docs.extend(self._load_from_folder(folder=folder))
  File "/home/netoferraz/opensource/langchain-o365/langchain/document_loaders/onedrive.py", line 108, in _load_from_folder
    loader = OneDriveFileLoader(file=file)
  File "pydantic/main.py", line 339, in pydantic.main.BaseModel.__init__
  File "pydantic/main.py", line 1076, in pydantic.main.validate_model
  File "pydantic/fields.py", line 860, in pydantic.fields.ModelField.validate
pydantic.errors.ConfigError: field "file" not yet prepared so type is still a ForwardRef, you might need to call OneDriveFileLoader.update_forward_refs().
```

File: langchain/document_loaders/sharepoint_file.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    from O365.drive import File
+
+CHUNK_SIZE = 1024 * 1024 * 5
+
+
+class SharePointFileLoader(BaseLoader, BaseModel):
+    class Config:
+        arbitrary_types_allowed = True
+
+    def load(self, file: File) -> List[Document]:  # type: ignore
```
Comment (netoferraz) [inline langchain/document_loaders/sharepoint_file.py:22]: The same explanation I did on the `OneDriveFileLoader`.

**Review by netoferraz — COMMENTED** [review]
If you guys need additional context or explanation, just le me know.

**Review by hwchase17 — APPROVED** [review]

File: langchain/document_loaders/onedrive_file.py (hunk 5 above)
Comment (hwchase17) [reply langchain/document_loaders/onedrive_file.py:22]: this signature shouldn't be changed. i can work to undo this

**Review by hwchase17 — DISMISSED** [review]
signature of load should not be changed

File: langchain/document_loaders/onedrive_file.py (hunk 5 above)
Comment (netoferraz) [reply langchain/document_loaders/onedrive_file.py:22]: If you have any suggestions on how to turn out this error, you could share them here, and I could try to implement it.

**Comment by eyurtsev** [comment]
@netoferraz Thank you for the contribution! As a heads up I'm in the process of adding a few abstractions to the document flow to decouple loading from parsing code. 

General strategy is here: 
https://github.com/hwchase17/langchain/pull/2833#issuecomment-1509097756

TLDR;  If you're able to implement a `BlobLoader` (interface and file system implementation shown below), it'll make it easier for users to add arbitrary parsers on top the loading interface.

```python
class BlobLoader(ABC):
    """Abstract interface for blob loaders implementation.

    Implementer should be able to load raw content from a storage system according
    to some criteria and return the raw content lazily as a stream of blobs.
    """

    @abstractmethod
    def yield_blobs(
        self,
    ) -> Iterable[Blob]:
        """A lazy loader for raw data represented by LangChain's Blob object.

        Returns:
            A generator over blobs
        """
```

Example:

https://github.com/hwchase17/langchain/blob/master/langchain/document_loaders/blob_loaders/file_system.py#L39

**Comment by netoferraz** [comment]
> @netoferraz Thank you for the contribution! As a heads up I'm in the process of adding a few abstractions to the document flow to decouple loading from parsing code.
> 
> General strategy is here: [#2833 (comment)](https://github.com/hwchase17/langchain/pull/2833#issuecomment-1509097756)
> 
> TLDR; If you're able to implement a `BlobLoader` (interface and file system implementation shown below), it'll make it easier for users to add arbitrary parsers on top the loading interface.
> 
> ```python
> class BlobLoader(ABC):
>     """Abstract interface for blob loaders implementation.
> 
>     Implementer should be able to load raw content from a storage system according
>     to some criteria and return the raw content lazily as a stream of blobs.
>     """
> 
>     @abstractmethod
>     def yield_blobs(
>         self,
>     ) -> Iterable[Blob]:
>         """A lazy loader for raw data represented by LangChain's Blob object.
> 
>         Returns:
>             A generator over blobs
>         """
> ```
> 
> Implementation for local file system:
> 
> https://github.com/hwchase17/langchain/blob/master/langchain/document_loaders/blob_loaders/file_system.py#L39

Hi @eyurtsev ! I'll try to bring those concepts to this loader and implement it.

**Comment by netoferraz** [comment]
> > @netoferraz Thank you for the contribution! As a heads up I'm in the process of adding a few abstractions to the document flow to decouple loading from parsing code.
> > General strategy is here: [#2833 (comment)](https://github.com/hwchase17/langchain/pull/2833#issuecomment-1509097756)
> > TLDR; If you're able to implement a `BlobLoader` (interface and file system implementation shown below), it'll make it easier for users to add arbitrary parsers on top the loading interface.
> > ```python
> > class BlobLoader(ABC):
> >     """Abstract interface for blob loaders implementation.
> > 
> >     Implementer should be able to load raw content from a storage system according
> >     to some criteria and return the raw content lazily as a stream of blobs.
> >     """
> > 
> >     @abstractmethod
> >     def yield_blobs(
> >         self,
> >     ) -> Iterable[Blob]:
> >         """A lazy loader for raw data represented by LangChain's Blob object.
> > 
> >         Returns:
> >             A generator over blobs
> >         """
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Implementation for local file system:
> > https://github.com/hwchase17/langchain/blob/master/langchain/document_loaders/blob_loaders/file_system.py#L39
> 
> Hi @eyurtsev ! I'll try to bring those concepts to this loader and implement it.

Hey @eyurtsev and @hwchase17. Finally, I had a chance to work again on this PR.  I decoupled the loading and parser process using `FileSystemBlobLoader` and  `BaseBlobParser`. Could you guys review it ?

**Comment by HoiDam** [comment]
Are you able to solve the issue? =]

**Comment by willemmulder** [comment]
Did anyone have a chance to look at this? Would love to have this merged :-)

**Comment by guidorietbroek** [comment]
@netoferraz great initiative!

Is this loader adding metadata from SPO, it’s possible in the native module [O365](https://github.com/O365/python-o365) .

**Comment by baskaryan** [comment]
> Any thoughts on resurrecting this so it can be merged? I'd like to send a PR to improve some of the auth handling and didn't want to step on this because I'd also like Sharepoint support. Thanks!

yes! working to revive

**Comment by netoferraz** [comment]
> Any thoughts on resurrecting this so it can be merged? I'd like to send a PR to improve some of the auth handling and didn't want to step on this because I'd also like Sharepoint support. Thanks!

It would be great if we can move forward with this PR! I'm not sure if the maintainers have any intentions to accept this work.

**Comment by baskaryan** [comment]
updated, would love one more review @netoferraz and @eyurtsev!

File: libs/langchain/langchain/document_loaders/base_o365.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    token_path: FilePath = Path.home() / ".credentials" / "o365_token.txt"
+
+
+class _FileType(str, Enum):
+    DOC = "doc"
+    DOCX = "docx"
+    PDF = "pdf"
+
+
+def fetch_mime_types(file_types: Sequence[_FileType]) -> Dict[str, str]:
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/base_o365.py:46]: Could we replace this function with this dict?

```
EXTENSION_TO_MIMETYPE = {
    "doc": "application/msword",
    "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "pdf": "application/pdf",
}
```

File: libs/langchain/langchain/document_loaders/base_o365.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        elif file_type.value == "docx":
+            mime_types_mapping[
+                file_type.value
+            ] = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"  # noqa: E501
+        elif file_type.value == "pdf":
+            mime_types_mapping[file_type.value] = "application/pdf"
+    return mime_types_mapping
+
+
+class O365BaseLoader(BaseLoader, BaseModel):
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/base_o365.py:60]: Haven't worked with O365 before. What type of stuff is accessible except for sharepoint? (Trying to understand why inheritance is needed)

File: libs/langchain/langchain/document_loaders/sharepoint.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    document_library_id: str = Field(...)
+    """ The ID of the SharePoint document library to load data from."""
+    folder_path: Optional[str] = None
+    """ The path to the folder to load data from."""
+    object_ids: Optional[List[str]] = None
+    """ The IDs of the objects to load data from."""
+
+    @property
+    def _file_types(self) -> Sequence[_FileType]:
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/sharepoint.py:26]: The new blob loaders abstraction helps to prevent hard-coding knowledge of parsing into content fetching. This makes the loading code a lot easier to reuse.

The loader should take 2 attributes that should be part of the initializer. 

1. Blob parser
2. extensions

File: libs/langchain/langchain/document_loaders/sharepoint.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        """Return supported file types."""
+        return _FileType.DOC, _FileType.DOCX, _FileType.PDF
+
+    @property
+    def _scopes(self) -> List[str]:
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/sharepoint.py:31]: This feels like an attribute that should handled via root validator instead with a default assigned to ["sharepoint", "basic"] in the absence of a value being provided.

File: libs/langchain/langchain/document_loaders/parsers/msword.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
+from typing import Iterator
+
+from langchain.document_loaders.base import BaseBlobParser
+from langchain.document_loaders.blob_loaders import Blob
+from langchain.schema import Document
+
+
+class MsWordParser(BaseBlobParser):
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/parsers/msword.py:8]: (can be refactored later) could expose an init to take a parameter (e.g., 'unstructured-doc' or 'unstructured-docx') and set the parser based on that, it'll allow users to also swap out strategies for parsing doc and docx files

File: libs/langchain/langchain/document_loaders/sharepoint.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
+"""Loader that loads data from Sharepoint Document Library"""
+from __future__ import annotations
+
+from typing import Iterator, List, Optional, Sequence
+
+from langchain.docstore.document import Document
+from langchain.document_loaders.base_o365 import (
+    O365BaseLoader,
+    _FileType,
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/sharepoint.py:9]: import a private attribute

File: libs/langchain/langchain/document_loaders/sharepoint.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+    def lazy_load(self) -> Iterator[Document]:
+        """Load documents lazily. Use this when working at a large scale."""
+        try:
+            from O365.drive import Drive, Folder
+        except ImportError:
+            raise ImportError(
+                "O365 package not found, please install it with `pip install o365`"
+            )
+        drive = self._auth().storage().get_drive(self.document_library_id)
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/sharepoint.py:43]: could move data fetching to base o365, and replace the share point loader with the generic loader which will allow users to swap out parsing strategies

File: libs/langchain/langchain/document_loaders/base_o365.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+                file = drive.get_item(object_id)
+                if not file:
+                    logging.warning(
+                        "There isn't a file with"
+                        f"object_id {object_id} in drive {drive}."
+                    )
+                    continue
+                if file.is_file:
+                    if file.mime_type in list(file_mime_types.values()):
+                        file.download(to_path=temp_dir, chunk_size=self.chunk_size)
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/base_o365.py:136]: can the file be fetched to memory and then this can yield a Blob directly?

File: libs/langchain/langchain/document_loaders/base_o365.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+        """Return supported file types."""
+
+    @property
+    def _fetch_mime_types(self) -> Dict[str, str]:
+        """Return a dict of supported file types to corresponding mime types."""
+        return fetch_mime_types(self._file_types)
+
+    @property
+    @abstractmethod
+    def _scopes(self) -> List[str]:
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/base_o365.py:80]: Could we convert the scopes into attributes rather than properties? They can be set using root validators if using pydantic or via the init if using vanilla classes

File: libs/langchain/langchain/document_loaders/base_o365.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    settings: _O365Settings = Field(default_factory=_O365Settings)
+    """Settings for the Office365 API client."""
+    auth_with_token: bool = False
+    """Whether to authenticate with a token or not. Defaults to False."""
+    chunk_size: Union[int, str] = CHUNK_SIZE
+    """Number of bytes to retrieve from each api call to the server. int or 'auto'."""
+
+    @property
+    @abstractmethod
+    def _file_types(self) -> Sequence[_FileType]:
```
Comment (eyurtsev) [inline libs/langchain/langchain/document_loaders/base_o365.py:70]: @baskaryan this could be a good candidate of a class to rewrite using the blob generators

Single blob generator takes in its interface:

- auth or settings for auth
- scopes

filters:
- drive
- object ids and/or folder
- file extensions
- chunk_size -- for large file downloads
- max file size -- limit to file size
- show progress (bool) -- progress indicator

it yields Blobs

---

Then can be combined with Generic Loader which will allow decoupling the hard-coding between parsers and content fetching

**Comment by eyurtsev** [comment]
@netoferraz :wave:  thank you for the contribution! I left a few comments on the PR, overall looks good to me, so okay merging as is. (cc @baskaryan )

There's a BlobLoader abstraction in the codebase that would fit the requirements here pretty well with an implementation for the file system called `FileSystemBlobLoader` that can be replicated here. The way it would look would be to declare something like

O365BlobLoader, it will take a bunch of attribtues in the init like auth, and filters, and yield blobs. 

Then one could compose it with `GenericLoader` to apply any sort of parser to content that can be fetched from O365

Not a requirement for merging this PR as we can re-use the existing code at a later point. :)

**Comment by netoferraz** [comment]
> @netoferraz 👋 thank you for the contribution! I left a few comments on the PR, overall looks good to me, so okay merging as is. (cc @baskaryan )
> 
> There's a BlobLoader abstraction in the codebase that would fit the requirements here pretty well with an implementation for the file system called `FileSystemBlobLoader` that can be replicated here. The way it would look would be to declare something like
> 
> O365BlobLoader, it will take a bunch of attribtues in the init like auth, and filters, and yield blobs.
> 
> Then one could compose it with `GenericLoader` to apply any sort of parser to content that can be fetched from O365
> 
> Not a requirement for merging this PR as we can re-use the existing code at a later point. :)

Thank you, @eyurtsev ! @baskaryan If you understand that we need to do some additional work based on the @eyurtsev review, let me know, ok? Otherwise, seems we could move to approve this work.
